Citizen Mourn - #2713

bob0bob opened the pull request:
This is an enhancement to citizen when a death of another citizen in the colony happen.

When a citizen dies for any reason (not a guard), regular citizen, a chat message will go out letting the player no that tomorrow the colony will mourn the loss of their citizen. This event will happen for the entire day.
NO citizens, other than guards, will do anything. They will all be idle. They will hang around the town hall for the entire day like they are mourning the loss of another citizen. At the end of the day the game will go back to normal mode and people will work.

Each time a citizen dies, this even will be triggered. Hope this helps people killing off their citizen to get a better skilled on.

Closes #
Closes #
Closes #

# Changes proposed in this pull request:
-
-
-

Review please

bob0bob committed:
Citizen Mourn - #2713

This is an enhancement to citizen when a death of another citizen in the colony happen.

When a citizen dies for any reason (not a guard), regular citizen, a chat message will go out letting the player no that tomorrow the colony will mourn the loss of their citizen. This event will happen for the entire day.
NO citizens, other than guards, will do anything. They will all be idle. They will hang around the town hall for the entire day like they are mourning the loss of another citizen. At the end of the day the game will go back to normal mode and people will work.

Each time a citizen dies, this even will be triggered. Hope this helps people killing off their citizen to get a better skilled on.

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:
```diff
@@ -411,6 +416,10 @@ public void onDeath(@NotNull final DamageSource damageSource)
+            {
+                citizenColonyHandler.getColony().setNeedToMourn(true, citizenData.getName());
+            }
             citizenColonyHandler.getColony().getCitizenManager().removeCitizen(getCitizenData());
```
I think it would be better to just let them mourn when the player killed the entity, else people could be quite angry about how often their citizens don't work.

bob0bob replied:
@Raycoms That could be changed.   I was thinking more of needing guards to protect your colony and making citizens happy.  But if citizens are dying that often then it could need up making them mourn everyday.

Got to ask,  in my colony I might have a guard die here and their but a regular citizen hasn't died in the last 30 days on my game.    But I see people are killing there citizens all the time.  Didn't get that either.  I guess people want faster leveling up and just keeps routing citizens until they get a good skill level.

Also, see all the time "I fired him" because he was not working.  That would mean bug, or them not understanding.

Raycoms replied:
Can we remove the inWall deaths here again, we still are having them, else I think it looks good.
How long do they mourn?

bob0bob replied:
They mourn for 2,000 - 4000 ticks..  I reduced it down from all day to that.   It just a partial day and not all of them mourn for the same amount.

I'll have to check to see where we have to do that check and how we pass it around so the ondeath routine will not call mourn on those deaths, but it should be easy.

Raycoms commented:
yeah our citizen kill detection needs some improvement yet, but atm miner
and lumberjack sometimes die, and when barbarians enter quite often at
least one dies


2018-08-03 14:28 GMT-03:00 bob0bob <notifications@github.com>:

> *@bob0bob* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java
> <https://github.com/ldtteam/minecolonies/pull/2726#discussion_r207614340>:
>
> > @@ -411,6 +416,10 @@ public void onDeath(@NotNull final DamageSource damageSource)
>              citizenColonyHandler.getColony().decreaseOverallHappiness(penalty);
>              triggerDeathAchievement(damageSource, citizenJobHandler.getColonyJob());
>              citizenChatHandler.notifyDeath(damageSource);
> +            if (!(citizenJobHandler.getColonyJob() instanceof AbstractJobGuard))
> +            {
> +                citizenColonyHandler.getColony().setNeedToMourn(true, citizenData.getName());
> +            }
>              citizenColonyHandler.getColony().getCitizenManager().removeCitizen(getCitizenData());
>
> @Raycoms <https://github.com/Raycoms> That could be changed. I was
> thinking more of needing guards to protect your colony and making citizens
> happy. But if citizens are dying that often then it could need up making
> them mourn everyday.
>
> Got to ask, in my colony I might have a guard die here and their but a
> regular citizen hasn't died in the last 30 days on my game. But I see
> people are killing there citizens all the time. Didn't get that either. I
> guess people want faster leveling up and just keeps routing citizens until
> they get a good skill level.
>
> Also, see all the time "I fired him" because he was not working. That
> would mean bug, or them not understanding.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2726#discussion_r207614340>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y8RIGl2Aw-bJJVycMKxYc0oXwMjsks5uNIg6gaJpZM4VtmoZ>
> .
>

bob0bob commented:
@Raycoms Could you not hook onto the death call in minecraft to catch it no matter what.   Then you could check if that is a citizen and get the death of the citizen.

Raycoms commented:
We catch any death, but we can't be sure that the player was involved.

bob0bob commented:
our saying you only want to mourn if a player kills the citizen.   Yeah,  players don't kill citizens.  they push them into lava or spill lava on them to kill them so their happiness doesn't get hit.

Only time you would get that is on a server where another players attacks a colony.

This catch is any non-guard dies, they mourn for a period of time the next day.  if it is a raid and several die that day, it doesn't matter.  They will only mourn once for everyone that died on that day.

The mourn time is a random thing for each citizen.  Some mourn longer then others.

bob0bob commented:
But, you guys are more in touch with what the server and how people play.  If it should be changed just let me know.   I tried posting it to discord to ask for help to make sure I got it in away that works with the flow of the game.  I didn't get any response after a couple of days.  
Is there a better way to submit ideas?  I didn't see anything.

Raycoms commented:
Uhm, okay, how about we restrict things, like barbarian deaths -> No
mourning, miner dieing underground, no mourning and lumberjack dieing from
suffocation no mourning, and no guards death as well. So what's missing is
players killing citizens

2018-08-03 17:04 GMT-03:00 bob0bob <notifications@github.com>:

> But, you guys are more in touch with what the server and how people play.
> If it should be changed just let me know. I tried posting it to discord to
> ask for help to make sure I got it in away that works with the flow of the
> game. I didn't get any response after a couple of days.
> Is there a better way to submit ideas? I didn't see anything.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2726#issuecomment-410361926>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y9gRra_otR5zZ20Z5CfEoB0nILzmks5uNKzMgaJpZM4VtmoZ>
> .
>

bob0bob commented:
Opps.  God that would be funny the colony morning a barbarian.  I forgot a barbarians is spawned from a citizen class.  I would love to see everyones comments on that.

Ok. if miner, no mourn.
if lumberjack no mourn. (Can you tell if a lumberjack dies from suffocation)

It has no guards death already.

Raycoms commented:
You can tell, by the damage source in wall

2018-08-03 21:57 GMT-03:00 bob0bob <notifications@github.com>:

> Opps. God that would be funny the colony morning a barbarian. I forgot a
> barbarians is spawned from a citizen class. I would love to see everyones
> comments on that.
>
> Ok. if miner, no mourn.
> if lumberjack no mourn. (Can you tell if a lumberjack dies from
> suffocation)
>
> It has no guards death already.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2726#issuecomment-410410106>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9yxl_-YiRqDag2BDf4j0MGoKZyXudks5uNPF9gaJpZM4VtmoZ>
> .
>

bob0bob commented:
I've got that from a citizen in their house.   Why not check for that as a source and have the citizen move?  I bet you have tried that and their was a problem with that.  What kind of problem prevented citizen from moving.

I found out it was the bed.   The citizen sleeping it not really sleeping.   It appears the graphics just makes the person appear to be asleep but the game thinks they still take two blocks.  and the bookshelf killed them as the game doesn't react to a citizen being attacked.

The game is to hard coded to this is the status and I'm doing that status.
Should create a AItask and active is the target source is a wall.  So no matter what the status of the citizen is you take over and move the citizen.

bob0bob commented:
Does your pathing system take into effect to check the height of the citizen.   So if the citizen is two blocks high, the next position has 2 air blocks.  I would say no because you can put a bed under a book shelf and not have 2 blocks clear and it will kill the citizen.

This would be a whole in the pathing.  My guess,  I've never looked at MC pathing methods.  I'm way to new to Minecraft coding and MC to even bring any good ideas to this area.

Raycoms commented:
It's not a problem with pathing that they come out, pathing checks for two
air blocks, but teleporting out of bed sometimes bugs with this.
And yeah, stuck citizens stop working with a block in their head, but a
stuck lumberjack in a tree is definitely stuck,

2018-08-03 22:28 GMT-03:00 bob0bob <notifications@github.com>:

> Does your pathing system take into effect to check the height of the
> citizen. So if the citizen is two blocks high, the next position has 2 air
> blocks. I would say no because you can put a bed under a book shelf and not
> have 2 blocks clear and it will kill the citizen.
>
> This would be a whole in the pathing. My guess, I've never looked at MC
> pathing methods. I'm way to new to Minecraft coding and MC to even bring
> any good ideas to this area.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2726#issuecomment-410412439>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y3EEhpZWFh0oInNyjjtNL89zDDkQks5uNPi9gaJpZM4VtmoZ>
> .
>

Raycoms commented:
Probably would be good to allow people to turn off mourning via config
option as well, case someone doesn't like it.

2018-08-03 22:31 GMT-03:00 Ray Neiheiser <ray.neiheiser@gmail.com>:

> It's not a problem with pathing that they come out, pathing checks for two
> air blocks, but teleporting out of bed sometimes bugs with this.
> And yeah, stuck citizens stop working with a block in their head, but a
> stuck lumberjack in a tree is definitely stuck,
>
> 2018-08-03 22:28 GMT-03:00 bob0bob <notifications@github.com>:
>
>> Does your pathing system take into effect to check the height of the
>> citizen. So if the citizen is two blocks high, the next position has 2 air
>> blocks. I would say no because you can put a bed under a book shelf and not
>> have 2 blocks clear and it will kill the citizen.
>>
>> This would be a whole in the pathing. My guess, I've never looked at MC
>> pathing methods. I'm way to new to Minecraft coding and MC to even bring
>> any good ideas to this area.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ldtteam/minecolonies/pull/2726#issuecomment-410412439>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AGI9y3EEhpZWFh0oInNyjjtNL89zDDkQks5uNPi9gaJpZM4VtmoZ>
>> .
>>
>
>

bob0bob committed:
formatting changes.

bob0bob committed:
formatting changes

bob0bob committed:
tweaking the mourning time.

bob0bob commented:
What is this waiting for.

Nightenom commented:
For resolving conflicts for sure, I don't know whether Ray is still waiting for something else

bob0bob commented:
I bet it has conflicts now.  been here too long many changes probably conflict with it.  I'll look at it later.

bob0bob committed:
Merge remote-tracking branch 'origin/version/1.12' into feature/Citizen-Mourn---#2713

bob0bob commented:
updated to fix conflicts.

bob0bob commented:
conflict resolved.   It just doesn't like one function and I'm not able to change it and doesn't seem like it is needed.

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count |  New  | Old  | Diff  |
| ------ |:-----:|:-----:|:----:|:-----:|
| Total  | 40227 | 15801 | 8658 | +7143 |
| Errors |  597  |  32   |  19  |  +13  |
##### More Information:
- [Analysis - Full](https://teamcity.minecolonies.com/viewLog.html?buildId=16323&tab=Inspection&buildTypeId=bt88)

Raycoms commented:
okay perfect

Am Mo., 20. Aug. 2018 um 23:55 Uhr schrieb bob0bob <notifications@github.com
>:

> *@bob0bob* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java
> <https://github.com/ldtteam/minecolonies/pull/2726#discussion_r211465485>:
>
> > @@ -411,6 +416,10 @@ public void onDeath(@NotNull final DamageSource damageSource)
>              citizenColonyHandler.getColony().decreaseOverallHappiness(penalty);
>              triggerDeathAchievement(damageSource, citizenJobHandler.getColonyJob());
>              citizenChatHandler.notifyDeath(damageSource);
> +            if (!(citizenJobHandler.getColonyJob() instanceof AbstractJobGuard))
> +            {
> +                citizenColonyHandler.getColony().setNeedToMourn(true, citizenData.getName());
> +            }
>              citizenColonyHandler.getColony().getCitizenManager().removeCitizen(getCitizenData());
>
> They mourn for 2,000 - 4000 ticks.. I reduced it down from all day to
> that. It just a partial day and not all of them mourn for the same amount.
>
> I'll have to check to see where we have to do that check and how we pass
> it around so the ondeath routine will not call mourn on those deaths, but
> it should be easy.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2726#discussion_r211465485>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9yziUG3ZZ95I4jDOQmf2Y0bayF9rEks5uS3aigaJpZM4VtmoZ>
> .
>

Raycoms commented:
@bob0bob did you finish the in wall death mourn avoidance?
And there are 2 more conflicts again

bob0bob committed:
Merge remote-tracking branch 'origin/version/1.12' into feature/Citizen-Mourn---#2713

bob0bob committed:
Mourning update

prevent mourning when citizen is killed by a wall.  Issues in the code that can cause a citizen to die from being in the wall.

bob0bob commented:
Added a check to see if citizen was killed from in_wall and prevent mourning on those types of deaths.  This should resolve all issues related to this new feature.

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  |   0   |  0  |  0  |  0   |
| Errors |   0   |  0  |  0  |  0   |
##### More Information:

Raycoms commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/2726/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

Raycoms committed:
easy test fix
